### PR TITLE
Fixed several bugs in code

### PR DESCRIPTION
**Pull request by QianyangPeng**
Bug1: globals should (only) be able to be declared and referred to with an identifier.

e.g. we can declare `(global $someglobal (mut i32) (i32.const 0))`
       but we shouldn't do `(global 0 (mut i32) (i32.const 0))`

Also, we need to be able to write `global.set $someglobal` rather than having to write the index for them.

Bug2: Control sequence instructions should take `TypeDecls` rather than `VecType`.

i.e. all type declaration should be in the format of `(result ValType)`, rather than `VecType`.

Bug 3: The order of function parameters needs to be reversed when putting into locals.

e.g.:

```
    (func $sub (export "sub")
        (param i32 i32)
        (result i32)
        (local.get 0)
        (local.get 1)
        (i32.sub)
        (return)
    )
    (invoke "sub" (i32.const 12) (i32.const 5))
```

This should return `7`, rather than `-7`.

### CODE REVIEW

File: data.md [hunk 1]
```diff
@@ -102,6 +102,15 @@ We can append two `ValTypes`s with the `_+_` operator.
     rule (VT:ValType VTYPES) + VTYPES' => VT (VTYPES + VTYPES')
 ```
 
+Also we can reverse a `ValTypes` with `#revt`
+
+```k
+    syntax ValTypes ::= #revt ( ValTypes ) [function]
+ // -------------------------------------------------
+    rule #revt(.ValTypes) => .ValTypes
+    rule #revt(V VS)    => #revt(VS) + V .ValTypes
```
Comment (ehildenb) [inline data.md]: formatting

File: wasm.md [hunk 2]
```diff
@@ -562,30 +562,28 @@ Note that, unlike in the WebAssembly specification document, we do not need the
...
-    syntax FoldedInstr ::= "(" "if" VecType Instrs "(" "then" Instrs ")" ")"
-                         | "(" "if" VecType Instrs "(" "then" Instrs ")" "(" "else" Instrs ")" ")"
-                         | "(" "if"         Instrs "(" "then" Instrs ")" "(" "else" Instrs ")" ")"
-    syntax BlockInstr  ::= "if" VecType Instrs "else" Instrs "end"
-                         | "if" VecType Instrs               "end"
+    syntax FoldedInstr ::= "(" "if" TypeDecls Instrs "(" "then" Instrs ")" ")"
+                         | "(" "if" TypeDecls Instrs "(" "then" Instrs ")" "(" "else" Instrs ")" ")"
+    syntax BlockInstr  ::= "if" TypeDecls Instrs "else" Instrs "end"
+                         | "if" TypeDecls Instrs               "end"
  // --------------------------------------------------------------
```
Comment (ehildenb) [inline wasm.md]: formatting

File: wasm.md [hunk 3]
```diff
@@ -651,11 +649,13 @@ When globals are declared, they must also be given a constant initialization val
...
     syntax Defn       ::= GlobalDefn
-    syntax GlobalDefn ::= "(" "global"               TextGlobalType Instr ")"
-                        | "(" "global" TextFormatIdx TextGlobalType Instr ")"
+    syntax GlobalDefn ::= "(" "global"            TextGlobalType Instr ")"
+                        | "(" "global" Identifier TextGlobalType Instr ")"
                         |     "global" GlobalType
  // ---------------------------------------------
-    rule <k> ( global ID:TextFormatIdx TYP:TextGlobalType IS:Instr ) => ( global TYP IS ) ... </k>
+    rule asGMut ( (mut T:ValType ) ) => var   T
+    rule asGMut (      T:ValType   ) => const T
```
Comment (ehildenb) [inline wasm.md]: asGMut is already defined above, why are you redefining it here?

File: wasm.md [hunk 4]
```diff
@@ -920,9 +922,9 @@ Unlike labels, only one frame can be "broken" through at a time.
     syntax Instr ::= "(" "invoke" Int ")"
```
Comment (ehildenb) [inline wasm.md]: Put some documentation for `invoke` in the markdown above this. Explain why the double reverse is here.

File: wasm.md (hunk 3 above)
Comment (QianyangPeng) [reply wasm.md]: Well it should be some merging issue..

File: data.md [hunk 5]
```diff
@@ -246,6 +256,9 @@ Operator `_++_` implements an append operator for sort `ValStack`.
 
     rule #drop(.ValTypes,   VALSTACK)                       => VALSTACK
     rule #drop(TYPE VTYPES, < TYPE > VAL:Number : VALSTACK) => #drop(VTYPES, VALSTACK)
+
+    rule #revs(.ValStack) => .ValStack
+    rule #revs(V : VS)    => #revs(VS) ++ V : .ValStack
```
Comment (ehildenb) [inline data.md]: This is a very slow way to reverse the list, actually is quadratic. Can we make it tail-recursive instead?

```
syntax ValStack ::= #revs ( ValStack ) | #revs ( ValStack , ValStack ) [klabel(#revsAux)]
rule #revs(VS) => #revs(VS, .ValStack)
rule #revs(.ValStack, VS') => VS'
rule #revs(V : VS, VS') => #revs(VS, V : VS')
```

Then it's linear and won't blow up the memory usage of the OCaml and LLVM backend. Don't forget the extra klabel.

File: data.md [hunk 6]
```diff
@@ -102,6 +102,15 @@ We can append two `ValTypes`s with the `_+_` operator.
     rule (VT:ValType VTYPES) + VTYPES' => VT (VTYPES + VTYPES')
 ```
 
+Also we can reverse a `ValTypes` with `#revt`
+
+```k
+    syntax ValTypes ::= #revt ( ValTypes ) [function]
+ // -------------------------------------------------
+    rule #revt(.ValTypes) => .ValTypes
+    rule #revt(V VS)      => #revt(VS) + V .ValTypes
```
Comment (ehildenb) [inline data.md]: Same here, make a tail-recursive version instead.

File: wasm.md [hunk 7]
```diff
@@ -562,30 +562,28 @@ Note that, unlike in the WebAssembly specification document, we do not need the
...
+    syntax FoldedInstr ::= "(" "if" TypeDecls Instrs "(" "then" Instrs ")" ")"
+                         | "(" "if" TypeDecls Instrs "(" "then" Instrs ")" "(" "else" Instrs ")" ")"
+    syntax BlockInstr  ::= "if" TypeDecls Instrs "else" Instrs "end"
+                         | "if" TypeDecls Instrs               "end"
+ // ----------------------------------------------------------------
+    rule <k> ( if TDECLS C:Instrs ( then IS ) )              => C ~> ( if TDECLS IS else .Instrs end ) ... </k>
+    rule <k> ( if TDECLS C:Instrs ( then IS ) ( else IS' ) ) => C ~> ( if TDECLS IS else IS'     end ) ... </k>
+
+    rule <k> if TDECLS IS          end => if TDECLS IS else .Instrs end ... </k>
+    rule <k> if TDECLS IS else IS' end => IS  ~> label gatherTypes(result, TDECLS) { .Instrs } VALSTACK ... </k>
```
Comment (ehildenb) [inline wasm.md]: line up the `...` in these rules.

Comment (QianyangPeng) [reply wasm.md]: fixed as suggested

**Pull request merged**